.circleci/config.yml

```diff
@@ -363,12 +363,69 @@ commands:
       - matrix_notify_failure_unless_pr
 
   install_dependencies_osx:
+    # an extra cache is used to only save the flag that communicates whether the cache exists.
+    # if this flag was set (a cache exist) we will remove all files located in /usr/local & /opt/homebrew.
+    # with this simple trick restoring the cache is much faster. circleci is generating warning messages,
+    # if a file from the cache is overwriting an already existing file on disk.
+    # Restoring the cache is much faster, if removed all potentially conflicting files before.
     steps:
-      # FIXME: We used to cache dependencies on macOS but now it takes longer than just installing
-      # them each time. See https://github.com/ethereum/solidity/issues/12925.
+      - restore_cache:
+          keys:
+            - b_osx-dependencies.{{ arch }}{{ checksum ".circleci/osx_install_dependencies.sh" }}-flag
       - run:
-          name: Install build dependencies
-          command: .circleci/osx_install_dependencies.sh
+          name: Installing dependencies / Restoring dependency cache
+          command: |
+            if test -f ~/b_osx-dependencies-flag
+            then
+              echo "Dependency flag exists: removing /usr/local and /opt/homebrew directories. These directories will be restored from cache."
+
+              # circleci is providing the circleci cli tools via some kind of symlink magic.
+              # so we just save the original symlinks and restore them later.
+              CIRCLECI_CMD=`which circleci`
+              CIRCLECI_AGENT_CMD=`which circleci-agent`
+              cp "${CIRCLECI_CMD}" /tmp/circleci
+              cp "${CIRCLECI_AGENT_CMD}" /tmp/circleci-agent
+
+              # homebrew is installed in /usr/local on intel macs, but in /opt/homebrew on apple silicon.
+              if test -d /opt/homebrew
+              then
+                sudo rm -rf /opt/homebrew
+                sudo mkdir -p /opt/homebrew/bin
+                sudo chmod 777 /opt/homebrew
+                sudo chmod 777 /opt/homebrew/bin
+              fi
+              sudo rm -rf /usr/local/*
+              sudo mkdir -p /usr/local/bin
+              sudo chmod 777 /usr/local
+              sudo chmod 777 /usr/local/bin
```

Comment:
Why it need to be set to `777` and not `755`?

Reply:
I was just seeing strange errors when I was restoring the cache.. but you might be right, that `755` is sufficient

Reply:
It might be needed to have those set to `777` before restoring the cache.. lets see

Reply:
ok its needed before the cache gets restored:
```
Skipping writing "usr/local/Cellar/boost/1.83.0/include/boost/align/detail/alignment_of_gcc.hpp" - mkdir /usr/local/Cellar: permission denied
Skipping writing "usr/local/Cellar/boost/1.83.0/include/boost/align/detail/alignment_of_msvc.hpp" - mkdir /usr/local/Cellar: permission denied
Skipping writing "usr/local/Cellar/boost/1.83.0/include/boost/align/detail/assume_aligned.hpp" - mkdir /usr/local/Cellar: permission denied
Skipping writing "usr/local/Cellar/boost/1.83.0/include/boost/align/detail/assume_aligned_clang.hpp" - mkdir /usr/local/Cellar: permission denied
Skipping writing "usr/local/Cellar/boost/1.83.0/include/boost/align/detail/assume_aligned_gcc.hpp" - mkdir /usr/local/Cellar: permission denied
Skipping writing "usr/local/Cellar/boost/1.83.0/include/boost/align/detail/assume_aligned_intel.hpp" - mkdir /usr/local/Cellar: permission denied
Skipping writing "usr/local/Cellar/boost/1.83.0/include/boost/align/detail/assume_aligned_msvc.hpp" - mkdir /usr/local/Cellar: permission denied
Skipping writing "usr/local/Cellar/boost/1.83.0/include/boost/align/detail/element_type.hpp" - mkdir /usr/local/Cellar: permission denied
Skipping writing "usr/local/Cellar/boost/1.83.0/include/boost/align/detail/integral_constant.hpp" - mkdir /usr/local/Cellar: permission denied
Skipping writing "usr/local/Cellar/boost/1.83.0/include/boost/align/detail/is_aligned.hpp" - mkdir /usr/local/Cellar: permission denied
```

Reply:
but youre right that its better to have those not set to `777` after the cache restore

Reply:
the cache restore is not executed as root, an thus mean that with `755` the directory is not writable for others (owner is root here). however, I agree to let the CI script as simple as possible. I will remove it.

Reply:
ok.. just saw that you pushed some commits